util/ranger/detacher.go

```diff
@@ -425,3 +425,35 @@ func DetachCondsForColumn(sctx sessionctx.Context, conds []expression.Expression
 	}
 	return detachColumnCNFConditions(sctx, conds, checker)
 }
+
+// MergeDNFItems4Col receives a slice of DNF conditions, merges some of them which can be built into ranges on a single column, then returns.
+// For example, [a > 5, b > 6, c > 7, a = 1, b > 3] will become [a > 5 or a = 1, b > 6 or b > 3, c > 7].
+func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression) []expression.Expression {
+	mergedDNFItems := make([]expression.Expression, 0, len(dnfItems))
+	col2DNFItems := make(map[int64][]expression.Expression)
+	for _, dnfItem := range dnfItems {
+		cols := expression.ExtractColumns(dnfItem)
+		// If this condition contains multiple columns, we can't merge it.
+		if len(cols) != 1 {
+			mergedDNFItems = append(mergedDNFItems, dnfItem)
+			continue
+		}
+
+		uniqueID := cols[0].UniqueID
+		checker := &conditionChecker{
+			colUniqueID: uniqueID,
+			length:      types.UnspecifiedLength,
+		}
+		// If we can't use this condition to build range, we can't merge it.
+		if !checker.check(dnfItem) {
```

Comment:
This check is not necessary since we have called `ExtractColumns` and checked it only contains one column?

Reply:
Because a condition only containing one column doesn't make sure this condition can be built into ranges, such as functions like `ifnull`. If we merged this kinds of conditions, `Selectivity` can't handle this DNF and will result in an infinite loop.